lib/rubocop/cop/sorbet/no_dynamic_contract_includes.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+#
+# In preparation for adding Sorbet to the codebase we must remove all dynamic imports as these
+# make type checking impossible. The biggest single offender in our codebase is `Dry::Types.module`
+# which includes all the types like Strict::String, Coercible::Int, Types::Bool, etc.
+#
+# Instead of using these auto-magic includes, we must use ths static path. In most cases this is
+# can be done automatically by this class, so:
+#
+# Strict::String -> Dry::Types['strict.string']
+# Coercible::Int -> Dry::Types['coercible.string']
+# Types::Bool -> Dry::Types['bool']
+# Strict::Array.of(Strict::String) -> Dry::Types["strict.array"].of
+# Strict::DateTime -> Dry::Types["strict.date_time"]
+# Instance(Carrier) -> Dry::Types::Definition.new(Carrier).constrained(type: Carrier)
+#
+# Some were not automate-able and must be done manually. These are:
+#
+# Any -> Dry::Types::Any
+# Hash -> Dry::Types["hash"]
+# Nil -> Dry::Types["nil"]
+#
+module RuboCop
+  module Cop
+    module Sorbet
+      class NoDynamicContractIncludes < Cop
+        MSG = "Sorbet disallows dynamic includes. Do not include Dry::Types.module directly; use direct path instead."
+        MSG_PATH = "Sorbet disallows dynamic includes. Use full Dry::Types path instead."
+
+        def_node_matcher :include_statement, <<-PATTERN
+          (send _ :include (send (const (const _ :Dry) :Types) _))
+        PATTERN
+
+        def_node_matcher :instance_statement, <<-PATTERN
+          (send _ :Instance (const ...))
+        PATTERN
+
+        def_node_matcher :const_statement, <<-PATTERN
+          (const (const _ {:Strict :Coercible}) _)
+        PATTERN
+
+        def_node_matcher :types_statement, <<-PATTERN
+          (const (const _ :Types) {:DateTime :String :Bool :Array :Int :Hash :Date})
+        PATTERN
+
+        def on_const(node)
+          types_statement(node) do |_statement|
+            add_offense(node, message: MSG_PATH)
+          end
+          const_statement(node) do |_statement|
+            add_offense(node, message: MSG_PATH)
+          end
+        end
+
+        def on_send(node)
+          instance_statement(node) do |_statement|
+            add_offense(node, message: MSG)
+          end
+          include_statement(node) do |_statement|
+            add_offense(node, message: MSG)
+          end
+        end
+
+        DYNAMIC_REFERENCE_PATTERN = NodePattern.new("$(const (const _ ${:Strict :Coercible}) $_)")
```

Comment:
Why do we need both that and the `def_node_matcher`s?

Reply:
Actually, the node matchers don't seem used anywhere?

Reply:
Good call; was able to re-use the `NodePattern`s defined here in the `def_node_matcher` statements.

The `def_node_matcher`s are needed in order to make the calls in `on_const` and `on_send` work

Reply:
https://github.com/Shopify/rubocop-sorbet/pull/27/files#diff-4892c3bddcbe192b3a1dd30691e5d63eR42